### DISCUSSION TITLE
Need help with MultiPort

### DISCUSSION THREAD

**Discussion by autosavemidi** [mido/mido#326] [Q&A] [answered]
I am working in a program that autosaves MIDI messages as they are played, and everything is working really well when using a single input port.
I want to extend to use 2 or 3 simultaneous ports (the app is aimed at VST piano players, and some use the piano as one MIDI device, and the sustain pedal as a second device in another USB slot). I was able to get it to work using two separate input ports, and then using 2 iter_pending() loops. The basic code is something like this:

ports = mido.get_input_names()
port_A = mido.open_input(ports[0])
port_B = mido.open_input(ports[1])

for msg_A in port_A.iter_pending():
    ...
for msg_B in port_B.iter_pending():
    ...

The problem with this is that the loops are sequential. When I test by sending the same midi message through the 2 ports, the same event when entering through port_B is processed with a few milliseconds of delay.

I want to test if the mido.ports.MultiPort() helps. My problem is that I cannot make it to work.

If I try:

multi = mido.ports.MultiPort([port_A, port_B])
for msg in multi:
    print (msg)

as suggested in the documentation, no  msg is printed...

I have done all sorts of syntax combinations for the ports in multi, but nothing seems to work.

EDIT:

I posted the same question at stackoverflow and received an helpful suggestion as a workaround:

open_ports = [port_A, port_B]
for port in open_ports:
        msg = port.poll()

Now the same MIDI message sent through both ports arrive simultaneously.

**Comment by rdoursenaud** [comment, marked as answer]
Thanks for sharing your solution.

I guess one could also use *callbacks* instead of *polling* depending on the requirements.
This would lighten the loop load at the cost of having to handle thread synchronisation according to [the docs](https://mido.readthedocs.io/en/latest/ports.html?highlight=callback#callbacks).
